Kernel/Devices/Audio/IntelHDA/Stream.cpp

```diff
@@ -0,0 +1,245 @@
+
+    TRY(m_buffers.with([&](auto& buffers) -> ErrorOr<void> {
+        buffers = TRY(MM.allocate_dma_buffer_pages(cyclic_buffer_size_in_bytes, "IntelHDA Stream Buffers"sv, Memory::Region::Access::ReadWrite));
+
+        // 3.3.38 Input/Output/Bidirectional Stream Descriptor Cyclic Buffer Length
+        m_stream_io_window->write32(StreamRegisterOffset::CyclicBufferLength, buffers->size());
+
+        // 3.3.39: Input/Output/Bidirectional Stream Descriptor Last Valid Index
+        VERIFY(number_of_buffers_required_for_cyclic_buffer_size <= 256);
+        m_stream_io_window->write16(StreamRegisterOffset::LastValidIndex, number_of_buffers_required_for_cyclic_buffer_size - 1);
+
+        // 3.6.2: Buffer Descriptor List
+        m_buffer_descriptor_list = TRY(MM.allocate_dma_buffer_page("IntelHDA Stream BDL"sv, Memory::Region::Access::ReadWrite));
+        auto bdl_physical_address = m_buffer_descriptor_list->physical_page(0)->paddr().get();
+        m_stream_io_window->write32(StreamRegisterOffset::BDLLowerBaseAddress, bdl_physical_address & 0xffffffffu);
+        m_stream_io_window->write32(StreamRegisterOffset::BDLUpperBaseAddress, bdl_physical_address >> 32);
+
+        // 3.6.3: Buffer Descriptor List Entry
+        auto* buffer_descriptor_entry = m_buffer_descriptor_list->vaddr().as_ptr();
+        for (u8 buffer_index = 0; buffer_index < buffers->page_count(); ++buffer_index) {
+            auto* entry = buffer_descriptor_entry + buffer_index * 0x10;
+            *reinterpret_cast<u64*>(entry) = buffers->physical_page(buffer_index)->paddr().get();
+            *reinterpret_cast<u32*>(entry + 8) = PAGE_SIZE;
+            *reinterpret_cast<u32*>(entry + 12) = 0;
+        }
+        return {};
+    }));
+    return {};
+}
+
+void Stream::reset()
+{
+    // 3.3.35: Input/Output/Bidirectional Stream Descriptor Control
+    if (m_running)
+        stop();
+
+    // Writing a 1 causes the corresponding stream to be reset. The Stream Descriptor registers
+    // (except the SRST bit itself), FIFO's, and cadence generator for the corresponding stream
+    // are reset.
+    auto control = read_control();
+    control |= StreamControlFlag::StreamReset;
+    write_control(control);
+
+    // After the stream hardware has completed sequencing into the reset state, it will report a
+    // 1 in this bit. Software must read a 1 from this bit to verify that the stream is in reset.
+    do {
+        microseconds_delay(25);
+        control = read_control();
+    } while ((control & StreamControlFlag::StreamReset) == 0);
+
+    // Writing a 0 causes the corresponding stream to exit reset.
+    control &= ~StreamControlFlag::StreamReset;
+    write_control(control);
+
+    // When the stream hardware is ready to begin operation, it will report a 0 in this bit.
+    // Software must read a 0 from this bit before accessing any of the stream registers
+    do {
+        microseconds_delay(25);
```

Comment:
Why do you delay read by 25 microseconds here? a comment on this would be very nice.

Reply:
It's an arbitrary delay value for a tight loop. Replaced it with an invocation to `wait_until`.